src/lib/tls/asio/asio_stream.h

```diff
@@ -335,22 +328,66 @@ class Stream
          boost::asio::detail::throw_error(ec, "shutdown");
          }
 
+   private:
+      /**
+       * @brief Internal wrapper type to adapt the expected signature of `async_shutdown`
+       *        to the completion handler signature of `AsyncWriteOperation`.
+       *
+       * This is boilerplate to ignore the `size_t` parameter that is passed to the
+       * completion handler of `AsyncWriteOperation`.
+       *
+       * @todo in C++14 and above this could be implemented as a mutable lambda expression
+       *       that captures `handler` by perfect forwarding, like so:
+       *
+       *       [h = std::forward<Handler>(handler)] (...) mutable { return h(ec); }
+       */
+      template <typename Handler>
+      class Wrapper
+         {
+         public:
+            Wrapper(Handler&& handler) : _handler(std::forward<Handler>(handler)) {}
+            void operator()(boost::system::error_code ec, size_t)
+               {
+               if(!ec)
```

Comment:
Alternatively, we could try to check `native_handle()->is_closed()` in `AsyncWriteOperation`.

Reply:
We decided to get rid of it.